src/server/transport/socketio-simultaneous.test.ts

```diff
@@ -0,0 +1,280 @@
+import { InitializeGame } from '../../core/initialize';
+import { PlayerView } from '../../core/player-view';
+import { _ClientImpl } from '../../client/client';
+import * as Masters from '../../master/master';
+import { Ctx } from '../../types';
+
+type SocketIOTestAdapterOpts = SocketOpts & {
+  clientInfo?: Map<any, any>;
+  roomInfo?: Map<any, any>;
+};
+
+class SocketIOTestAdapter extends SocketIO {
+  constructor({
+    clientInfo = new Map(),
+    roomInfo = new Map(),
+    ...args
+  }: SocketIOTestAdapterOpts = {}) {
+    super(Object.keys(args).length ? args : undefined);
+    this.clientInfo = clientInfo;
+    this.roomInfo = roomInfo;
+  }
+
+  public get getPerMatchQueue() {
+    return this.perMatchQueue;
+  }
+}
+
+jest.mock('koa-socket-2', () => {
+  class MockSocket {
+    id: string;
+    callbacks: {};
+    emit: jest.Mock<any, any>;
+    broadcast: { emit: jest.Mock<any, any> };
+
+    constructor() {
+      this.id = 'id';
+      this.callbacks = {};
+      this.emit = jest.fn();
+      this.broadcast = { emit: jest.fn() };
+    }
+
+    async receive(type, ...args) {
+      await this.callbacks[type](args[0], args[1], args[2], args[3], args[4]);
+      return;
+    }
+
+    on(type, callback) {
+      this.callbacks[type] = callback;
+    }
+
+    to() {
+      return {
+        broadcast: this.broadcast,
+        emit: this.emit,
+      };
+    }
+
+    join() {}
+  }
+
+  class MockIO {
+    socket: MockSocket;
+    socketAdapter: any;
+
+    constructor() {
+      this.socket = new MockSocket();
+    }
+
+    adapter(socketAdapter) {
+      this.socketAdapter = socketAdapter;
+    }
+
+    attach(app) {
+      app.io = app._io = this;
+    }
+
+    of() {
+      return this;
+    }
+
+    on(type, callback) {
+      callback(this.socket);
+    }
+  }
+
+  return MockIO;
+});
+
+describe('simultaneous moves on server game', () => {
+  const game = {
+    name: 'test',
+    setup: ctx => {
+      const G = {
+        players: {
+          '0': {
+            cards: ['card3'],
+          },
+          '1': {
+            cards: [],
+          },
+        },
+        cards: ['card0', 'card1', 'card2'],
+        discardedCards: [],
+      };
+      return G;
+    },
+    playerView: PlayerView.STRIP_SECRETS,
+    turn: {
+      activePlayers: { currentPlayer: { stage: 'A' } },
+      stages: {
+        A: {
+          moves: {
+            A: {
+              client: false,
+              move: (G, ctx: Ctx) => {
+                const card = G.players[ctx.playerID].cards.shift();
+                G.discardedCards.push(card);
+              },
+            },
+            B: {
+              client: false,
+              ignoreStaleStateID: true,
+              move: (G, ctx: Ctx) => {
+                const card = G.cards.pop();
+                G.players[ctx.playerID].cards.push(card);
+              },
+            },
+          },
+        },
+      },
+    },
+  };
+  let db = new InMemory();
```

Comment:
I might be missing something here, but shouldn’t we use asynchronous storage to test this? Or run tests with both a sync and async storage implementation?

Reply:
Yeap, I can do that.

Reply:
Hi, @delucis . I used the last days to fix some bugs in my app and today I made some tests with thoses fixes. Using the PQueue without interval and InMemory storage, I played 4 matches in a row and no errors occurred.  :smile: 
That concludes that the errors I've reported were on the client, not in the engine.

Reply:
So, today I'm gonna push my project to the cloud and test it online. After that I will clean up the code removing unnecessary stuff and it will be ready to mere. Would you like me to test something else?

Reply:
Code pushed. Should I write a documentation for it, too?